### PR TITLE
Collapse if statements with comments

### PR DESCRIPTION
**Pull request by kgevorkyan**
### What's done:
* Change collapseThenBlocks function: KtBlockExpression.statements didn't hold comments, now it's fixed
* Monitor, which types of nodes are followed after nested if (more proper way), instead of check, that this if is the last child
* Fix indentions in tests

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt [hunk 1]
```diff
@@ -85,33 +84,46 @@ class CollapseIfStatementsRule(configRules: List<RulesConfig>) : DiktatRule(
...
             return null
         }
-        // We monitor which types of nodes are followed before nested `if`
+        // We monitor which types of nodes are followed before and after nested `if`
         // and we allow only a limited number of types to pass through.
         // Otherwise discovered `if` is not nested
         val listOfNodesBeforeNestedIf = parentThenNode.getChildren(null).takeWhile { it.elementType != IF }
-        val allowedTypes = listOf(LBRACE, WHITE_SPACE, RBRACE, KDOC, BLOCK_COMMENT, EOL_COMMENT)
-        if (listOfNodesBeforeNestedIf.any { it.elementType !in allowedTypes }) {
+        val listOfNodesAfterNestedIf = parentThenNode.getChildren(null).takeLastWhile { it != parentThenNode.findChildByType(IF) }
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt:96]: Why isn't `elementType` check enough here? Also, how will it work in case of several independent `if`s in the same block?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt [hunk 2]
```diff
@@ -85,33 +84,46 @@ class CollapseIfStatementsRule(configRules: List<RulesConfig>) : DiktatRule(
...
+        val allowedTypes = listOf(LBRACE, WHITE_SPACE, RBRACE, BLOCK_COMMENT, EOL_COMMENT)
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt:97]: Why did you remove kdoc?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt [hunk 3]
```diff
@@ -85,33 +84,46 @@ class CollapseIfStatementsRule(configRules: List<RulesConfig>) : DiktatRule(
...
     private fun collapse(parentNode: ASTNode, nestedNode: ASTNode) {
         collapseConditions(parentNode, nestedNode)
         collapseThenBlocks(parentNode, nestedNode)
     }
 
     private fun collapseConditions(parentNode: ASTNode, nestedNode: ASTNode) {
+        // If there are comments before nested if, we will move them into parent condition
+        val comments = takeCommentsBeforeNestedIf(parentNode)
+        val commentsText = if (comments.isNotEmpty()) {
+            "\n${comments.joinToString("\n") { it.text }}\n"
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt:123]: You can use `prefix` and `suffix` arguments of `joinToString` (I may be wrong with their names but they are definitely avaialble)

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt [hunk 4]
```diff
@@ -140,12 +151,30 @@ class CollapseIfStatementsRule(configRules: List<RulesConfig>) : DiktatRule(
...
+            }
+            parentNode.removeChild(it)
+        }
         // Merge parent and nested `THEN` blocks
         val nestedThenNode = (nestedNode.psi as KtIfExpression).then
-        val nestedThenText = (nestedThenNode as KtBlockExpression).statements.joinToString("\n") { it.text }
-        val newNestedNode = KotlinParser().createNode(nestedThenText)
+        val nestedContent = (nestedThenNode as KtBlockExpression).children().toMutableList()
+        // Remove {, }, and white spaces
+        repeat(2) {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt:167]: It's better to remove specific nodes eplicitly by type. Your algorithm will fail for block like
```kotlin
if (foo) {bar()}
```
without any spaces

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt (hunk 2 above)
Comment (kgevorkyan) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt:97]: Usually kdoc is used before classes or functions. Seems, that to use kdoc in `if` body - isn't good practice. Moreover, even KotlinParser failed with error, while takes the code, which have kdoc in if-statements.
For those reasons I decide, that we may not expect kdoc in if-statements

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt (hunk 1 above)
Comment (kgevorkyan) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt:96]: Previously, there were the check for nested if: we assumed, that current `if-statement` is nested, if **it is** the last child for external `if (not considered whitespaces and brackets)

Now, by analogy `listOfNodesBeforeNestedIf,` we suppose, that nested if not only the last child, but more - after nested `if` it also could be comments
In this case we can collapse statements too

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt (hunk 2 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt:97]: It sure is a bad practice, but it's really strange that it causes errors. I thought that in cases like here, when no special meaning of kdoc is needed, it is treated as an ordinary comment. Please at least add a comment describing why we don't check kdocs

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt (hunk 1 above)
Comment (kgevorkyan) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt:96]: If there will be several independent if-statements, like this
```kotlin
if (cond1) {
    if (cond2) {
    }
    if (cond3) {
    }
}
```
we won't collapse anything, since `if (cond2)` for the first look seems to be nested, but after this node we can see another IF,
which is not any type of `LBRACE, WHITE_SPACE, RBRACE, BLOCK_COMMENT, EOL_COMMENT`
And it's correct behavior, since we shouldn't collapse it like this:
```kotlin
if (cond1 && cond2) {
    if (cond3) {
    }
}
```

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt (hunk 3 above)
Comment (kgevorkyan) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt]: Done

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt (hunk 4 above)
Comment (kgevorkyan) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt:167]: I supposed, that we can assume, that work with the proper code style code (since it could be already fixed by diktat)
However, I agree, it's better to remove with knowledge about type of node. Done

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt (hunk 2 above)
Comment (kgevorkyan) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt:97]: What about creation of a some new rule, which will change the kdoc comments to the block comments, if they are in incorrect places? Like in the example, that we now discuss

Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt:97]: Good idea, we could add it to Chapter 2

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt (hunk 1 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt:99]: Yes, I remember that we discussed it, but I've mixed the details. Alright then

**Review by petertrr — APPROVED** [review]

**Comment by kgevorkyan** [comment]
@petertrr I improved the way of comments processing, check please last commits

**Review by petertrr — APPROVED** [review]
LGTM
